java/arcs/core/storage/WriteBack.kt

```diff
@@ -0,0 +1,145 @@
+
+    /** Await completion of all active flush jobs. */
+    suspend fun awaitIdle()
+}
+
+/** The factory interfaces of [WriteBack] implementations. */
+interface WriteBackFactory {
+    fun create(
+        protocol: String = "",
+        /**
+         * Provide a dedicated write-back thread pool or leave it to implementations
+         * to decide what threads to be designated as the write-backers.
+         */
+        writebackThreads: ExecutorService? = null,
+        /**
+         * The maximum queue size above which new incoming flush jobs will be suspended.
+         */
+        queueSize: Int = 10
+    ): WriteBack
+}
+
+/** Write-back implementation for Arcs Stores.*/
+class StoreWriteBack private constructor(
+    protocol: String,
+    writebackThreads: ExecutorService?,
+    queueSize: Int
+) : WriteBack,
+    Channel<suspend () -> Unit> by Channel(queueSize),
+    CoroutineScope by CoroutineScope(
+        writebackThreads?.asCoroutineDispatcher() ?: Dispatchers.IO
+    ),
+    Mutex by Mutex() {
+    // Only apply write-back to physical storage media(s).
+    private val passThrough = protocol != Protocols.DATABASE_DRIVER
+
+    // The number of active flush jobs.
+    private var activeJobs = 0
+
+    // The signal to block/release who are waiting for completion of active flush jobs.
+    private val awaitSignal = Mutex()
+
+    init {
+        // One of write-back thread(s) will wake up and execute flush jobs in FIFO order
+        // when there are pending flush jobs in queue. Powerful features like batching,
+        // merging, filtering, etc can be implemented at this call-site in the future.
+        if (!passThrough) {
+            launch {
+                try {
+                    while (true) {
+                        exitFlushSection { receive()() }
+                    }
+                } finally {
+                    if (awaitSignal.isLocked) {
+                        awaitSignal.unlock()
+                    }
+                }
+            }
+        }
+    }
+
+    override suspend fun flush(job: suspend () -> Unit) {
+        if (!passThrough) flushSection { job() }
+        else job()
+    }
+
+    override suspend fun asyncFlush(job: suspend () -> Unit) {
+        // Queue up a flush task can run average 3x-5x faster than launching it.
+        if (!passThrough) enterFlushSection { send(job) }
+        else job()
+    }
+
+    override suspend fun awaitIdle() {
+        if (!passThrough) awaitSignal.withLock {}
+    }
+
+    private suspend inline fun flushSection(job: () -> Unit) {
+        enterFlushSection()
+        job()
+        exitFlushSection()
+    }
+
+    private suspend inline fun enterFlushSection(job: () -> Unit = {}) {
+        withLock { if (++activeJobs == 1) awaitSignal.lock() }
+        job()
+    }
+
+    private suspend inline fun exitFlushSection(job: () -> Unit = {}) {
+        job()
+        withLock { if (--activeJobs == 0) awaitSignal.unlock() }
+    }
```

Comment:
Could this just be done with a semaphore?

Reply:
Ah it doesn't and shouldn't act like a semaphore because the number of permits is not fixed.
The correct semantics here is the very first IN holds the lock and the very last OUT releases the lock. So there could be 1...N active flushers and we ensure the first flusher claims busyness and the last flusher claims idleness.